src/librustc_trait_selection/traits/error_reporting/mod.rs

```diff
@@ -404,6 +409,20 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                             err.emit();
                             return;
                         }
+                        if is_unsize {
```

Comment:
This code seems fine, though I'm not sure if it's the best approach, versus modifying the tips below not to fire. I guess it seems fine. 

Nit: I'd put a blank line before the `if`

Reply:
I think I agree - I've moved the if-guard onto the tip that's causing the ICE and updated the comments

Reply:
This added an extra note to the diagnostic (which I think is a good thing) and I've updated the test accordingly